Autodetect asm.cpu whenever possible

valdaarhun opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

get_cpu_arm() scans through the ".ARM.attributes" section of the ELF file to find the "aeabi" subsection. This subsection contains general build attributes that records data about the compatibility of the ELF file. The relevant tags in this subsection (eg: Tag_CPU_arch) can help determine `asm.cpu`.

References:
- https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst
- https://github.com/ARM-software/abi-aa/blob/main/addenda32/addenda32.rst

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Will think of one after finishing the implementation.

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

Fixes #3747

valdaarhun committed:
ARM: Implement search for public (aeabi) attributes subsection

get_cpu_arm() scans through the ".ARM.attributes" section of the
ELF file to find the "aeabi" subsection. This subsection contains
general build attributes that records data about the compatibility
of the ELF file.

XVilka reviewed (changes requested):
Yes, this direction is correct. Once you finish it, please also add a test.

valdaarhun commented:
> Yes, this direction is correct. Once you finish it, please also add a test.

Sure,  I'll do that.

XVilka commented:
@valdaarhun do you have any progress on this? It would be nice to include it into the upcoming 0.7.0 release

valdaarhun commented:
> @valdaarhun do you have any progress on this?

Hi. Yes, I have made some progress. There's not much left.

> It would be nice to include it into the upcoming 0.7.0 release

Got it. I should be able to finish the implementation soon.

valdaarhun committed:
ARM: Add macros defining tags in the public attributes subsection

Each tag's value is defined as a null-terminated string or a
uleb128-encoded number. This info will be required while
searching for "TAG_CPU_ARCH" and "Tag_CPU_ARCH_PROFILE" in
the "ARM_TAG_FILE" sub-subsection.

valdaarhun committed:
ARM: Parse aeabi subsection

Rename "read_arm_attributes_section()" to "read_arm_aeabi_section()".
This function searches for the "TAG_FILE" sub-subsection.

valdaarhun commented:
TODO:

- [ ] Parse tags correctly in `read_arm_build_attributes()`. Some tags such as `TAG_COMPATIBILITY` have a few quirks but handling them should not be too hard.
- [ ] Testing

valdaarhun committed:
ARM: Add macros for ARM CPU arch version and profile

XVilka commented on librz/bin/format/elf/elf_info.c:
```diff
@@ -845,6 +845,75 @@ static char *get_cpu_mips(ELFOBJ *bin) {
+
+static char *read_arm_aeabi_section(char *ptr, ut32 bytes_to_read, bool isbig) {
+	while (bytes_to_read > 0) {
+		ut8 sub_subsection = *ptr;
```
use `rz_read_ble8()` for consistency.

valdaarhun replied:
Got it. Thanks for pointing this out.

DMaroo commented on librz/bin/format/elf/elf_info.c:
```diff
@@ -845,6 +845,75 @@ static char *get_cpu_mips(ELFOBJ *bin) {
+		}
+		subsection_ptr += subsection_size;
+	}
+
```
Minor nitpick: Put a goto label here, and jump here in all the error cases rather than returning the same string at multiple places.

valdaarhun replied:
Hi. Thanks for the review. I don't think this is required here since it's only returning one thing and not doing anything else. In the source code, it seems like `goto` is used when several things are required to be done in that particular control path (for eg: cleanup tasks before returing a value).

More info [here](https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md?plain=1#L78).

DMaroo replied:
Totally up to you, but my reasoning behind the suggestion was to make it easier to reason about the code. Am I sure that the _same_ string is returned everywhere? If we change it in the future, can I be sure I covered all the conditions?

DMaroo commented on librz/bin/format/elf/glibc_elf.h:
```diff
@@ -2712,6 +2712,107 @@ enum {
+#define TAG_CPU_RAW_NAME         4
+#define TAG_CPU_NAME             5
+#define TAG_CPU_ARCH             6
+#define Tag_CPU_ARCH_PROFILE     7
```
```suggestion
#define TAG_CPU_ARCH_PROFILE     7
```

valdaarhun replied:
Nice catch! Copypasta formatting error. I'll fix this.

valdaarhun committed:
Minor changes in ARM tags

XVilka commented:
@valdaarhun hi, have you had any progress on this PR?

valdaarhun commented:
> @valdaarhun hi, have you had any progress on this PR?

Hi. Sorry, I wasn't able to give much time to this last week. I do have some more commits to push. I'll push those for review. There's some more implementation left to do in the parser. I'll finish that up as well.

valdaarhun committed:
ARM: Read arm build attributes

* Read arm build attributes
* Incorporate reviewed changes

valdaarhun committed:
ARM: Add structure for arm cpu translation

valdaarhun committed:
ARM: Get arch name

XVilka reviewed (changes requested):
It breaks DWARF profile loading in some of the tests in its current state:
```
[XX] db/cmd/dwarf "ARM types"
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'aaa
afvl @ dbg.main
pdf @ dbg.new_some
aaa
' bins/elf/test_app2/test_app2.elf
-- stdout
--- expected
+++ actual
@@ -4,15 +4,15 @@
 var int32_t var_2ch @ stack - 0x2c
 var int32_t var_1ch @ stack - 0x1c
 var some_t *s @ stack - 0x1a
-arg int argc @ r0
-arg char **argv @ r1
-var struct Some *gg @ r2
+arg int argc @ reg0
+arg char **argv @ reg1
+var struct Some *gg @ reg2
 var float a @ ...
 var float b @ ...
 var double c @ ...
             ; CALL XREFS from dbg.main @ 0x817c, 0x8184
 / some_t * new_some()
-|           ; var struct Some *n @ r4
+|           ; var struct Some *n @ reg4
```

valdaarhun commented:
> It breaks DWARF profile loading in some of the tests in its current state:
> 
> ```
> [XX] db/cmd/dwarf "ARM types"
> RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'aaa
> afvl @ dbg.main
> pdf @ dbg.new_some
> aaa
> ' bins/elf/test_app2/test_app2.elf
> -- stdout
> --- expected
> +++ actual
> @@ -4,15 +4,15 @@
>  var int32_t var_2ch @ stack - 0x2c
>  var int32_t var_1ch @ stack - 0x1c
>  var some_t *s @ stack - 0x1a
> -arg int argc @ r0
> -arg char **argv @ r1
> -var struct Some *gg @ r2
> +arg int argc @ reg0
> +arg char **argv @ reg1
> +var struct Some *gg @ reg2
>  var float a @ ...
>  var float b @ ...
>  var double c @ ...
>              ; CALL XREFS from dbg.main @ 0x817c, 0x8184
>  / some_t * new_some()
> -|           ; var struct Some *n @ r4
> +|           ; var struct Some *n @ reg4
> ```

I am not sure why this is happening. I'll investigate this further.

XVilka commented:
@wargio, we will need the same approach for MIPS files. Apparently, corresponding ELF files contain more precise information about the CPU used than we currently process.

wargio commented:
This should be done with rz-arch. once we have it, we can define the CPU directly (or features) and just follow the same strategy for the recent mips PR.
